lib/find-python.js

```diff
@@ -1,13 +1,26 @@
 'use strict'
 
-const path = require('path')
 const log = require('npmlog')
 const semver = require('semver')
 const cp = require('child_process')
 const extend = require('util')._extend // eslint-disable-line
 const win = process.platform === 'win32'
 const logWithPrefix = require('./util').logWithPrefix
 
+const systemDrive = process.env.SystemDrive || 'C:'
+const username = process.env.USERNAME || process.env.USER
+const winDefaultLocationsArray = []
+
+for (const majorMinor of ['39', '38', '37', '36']) {
+  winDefaultLocationsArray.push(
```

Comment:
Instead of fixed `Users` and `Program Files` shouldn't this be with the environment variables `%ProgramFiles%`, `%ProgramFiles(x86)%`, `%ProgramW6432%` and `%LocalAppData%`?

Because these folders can be set to be on other disks than the system drive, and avoids potential localization issues in the path.

Reply:
@mfonville Thanks for the suggestion. I did implement your suggestion on another branch here: https://github.com/DeeDeeG/node-gyp/commit/e57fdcc5443e57be5c96107aab8ffafc3c2d5321

It does work. It's also a bit more involved. After some quick research and testing, I am somewhat doubtful whether there are circumstances where the paths actually change based on localization or user customization. See my thoughts below. If you have examples or more information about when/if this approach is necessary, please let me know.

At the moment I am inclined to believe this is not necessary. But I would not mind being proved wrong. Thanks.

---

My thoughts on this: I wish I knew in what circumstances the "Program Files" or "AppData" folders are localized or moved, if there are any such circumstances. I do not think the paths are truly localized on-disk, nor movable, in Windows 10. It's unclear to me if they've ever been truly localized on-disk, or movable in previous Windows releases.

_(Paths were slightly different in Windows XP, but Windows XP is not compatible with the latest `node-gyp`. Python 3.4 is the last version supporting Windows XP. `node-gyp` master branch now requires Python 3.6 and up, so no XP support anymore. And only Windows 10 is in "mainstream" support by Microsoft now.)_

Regarding custom locations for Program Files or AppData: Is anyone here aware of a proper way to move these folders? _(I am aware that folders such as Documents and Videos can be easily moved to another drive. But can "Program Files", user home folders, or "AppData" be moved off of the system drive? I have looked this up online, and the overall impression I am getting is that there is no official way to move these, and that moving them may break things.)_

Regarding localization, or Windows in languages other than English: I don't see the paths being actually localized on the disk. (On Windows 10 in Spanish, I see in the File Explorer, "C: > Usuarios > [user]`, but the moment you try to edit the path in the address bar, it changes to `C:\Users\[user]`. The env var `LOCALAPPDATA` is still `C:\Users\[user\\AppData\Local`. The env var `ProgramFiles` is still `C:\Program Files`. Even while this shows as "C:\Archivos de programa` in File Explorer, the moment you try to edit it, it changes to `C:\Program Files`.

Reply:
If XP is out of the picture, then I do think localization is probably indeed not an issue anymore. Windows 10 does this nicely.

But about custom paths, to be honest I am not working with Windows daily anymore (switched to Linux many years ago). But when a couple of years ago I did do some volunteering with managing some Windows systems, we could set custom locations for these paths through policies. And also I have seen customized paths when computers were upgraded from a HDD to a SSD, where these paths were customized to optimize performance.
I am not sure how Windows 10 handles it these days, but imho those environment variables are more reliable than any hardcoded path. So better to use the variables :-)

Reply:
Thanks for that explanation. I also found this official Microsoft document about moved user folders: https://docs.microsoft.com/en-us/troubleshoot/windows-server/user-profiles-and-logon/relocation-of-users-and-programdata-directories

So apparently user directories _can_ be moved to a custom location. (Maybe this applies only to Windows Server? I'd say `node-gyp` is not any less important for Windows Server users.)

I'm a little more inclined to do this (https://github.com/DeeDeeG/node-gyp/commit/e57fdcc5443e57be5c96107aab8ffafc3c2d5321) now. Truth be told, it's not much more complex. Given even a mild justification for end-user benefit, I'd want to do it.

Reply:
And most of all, there is no downside to the change I think.

Reply:
Go for it.  This code is easy to read/ understand.

Reply:
I know that is the Swiss in me but...

Please sort your const declarations in the order that they are used to improve readability.

Reply:
More research: We definitely don't have to worry about Windows XP or Vista.

Node v6 refuses to run on anything older than Windows 7. Node v5 is the newest Node that will run on Windows Vista. _Node v5 cannot even run the latest `node-gyp`_, as `node-gyp` user newer JavaScript features that aren't supported in Node 5. In conclusion: modern `node-gyp` won't run at all on Windows Vista, regardless of this PR.

Reply:
> environment variables [. . .] `%ProgramW6432%` [. . .]

@mfonville thanks for mentioning that. I believe I will need to switch from `process.env.ProgramFiles` to `process.env.ProgramW6432` to get the "not (x86)" Program Files.  

<details>

It appears we have to use `process.env.ProgramW6432` to get the not-architecture-specific `Program Files` location. From 32-bit Node on a 64-bit host, `process.env.ProgramFiles` returns `[driveLetter]:\Program Files (x86)`. As such, `ProgramW6432` is the only easy way to get `[driveLetter]:\Program Files` from the environment across all combinations of host arch/node arch.

(Official documentation says that the env var `ProgramW6432` only exists as of Windows 7 or newer, but that's okay. `node-gyp` doesn't even run on Windows Vista or older.)

</details>

See:
- https://docs.microsoft.com/en-us/windows/win32/winprog64/wow64-implementation-details?redirectedfrom=MSDN#environment-variables
- https://serverfault.com/questions/413320/programfiles-differences-on-64bit-windows

Reply:
Okay, I committed this in the PR now. `Program Files` and `AppData\Local` are set from environment variables now.